### PR TITLE
[iOS] Add new browser menu UI

### PR DESCRIPTION
**Pull request by kylehickinson**
This adds a new modern browser menu UI behind the feature flag `brave-use-modern-browser-menu`

Resolves https://github.com/brave/brave-browser/issues/42836

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Menu:
| Visible | Show All | Show All Scrolled | 
| --- | --- | --- | 
| default visible | default view all | default view all scrolled |

### Customize:
| Default | Scrolled | 
| --- | --- |
| customize | customize scrolled |

### CODE REVIEW

File: ios/brave-ios/Sources/BrowserMenu/BrowserMenuModel.swift [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    self.visibleActions = sortedActions
+    self.hiddenActions = hiddenActions
+  }
+
+  private func rank(for action: Action) -> Double {
+    actionRanks.value[action.id.id] ?? Double(action.id.defaultRank)
+  }
+
+  func updateActionVisibility(_ action: Action, visibility: Action.Visibility) {
+    actionVisibility.value[action.id.id] = visibility == .visible ? true : false
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/BrowserMenu/BrowserMenuModel.swift:89]: nit: `visibility == .visible`. No need for the `? true : false` :D

File: ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      "BrowserMenu.hiddenActionsTitle",
+      bundle: .module,
+      value: "Available Actions",
+      comment:
+        "A title shown above the list of actions that are hidden on the menu but available for the user to make visible."
+    )
+    public static let showAllButtonTitle = NSLocalizedString(
+      "BrowserMenu.showAllButtonTitle",
+      bundle: .module,
+      value: "Show All…",
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift:65]: nit: The `…` is a special Unicode character. It should be three separate dots like: `...`?

File: ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    public static let braveNews = NSLocalizedString(
+      "ActionTitles.braveNews",
+      bundle: .module,
+      value: "Brave News",
+      comment: "A button title shown on the menu that presents the Brave News feature."
+    )
+    public static let share = NSLocalizedString(
+      "ActionTitles.share",
+      bundle: .module,
+      value: "Share…",
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift:105]: nit: The `…` is a special unicode character. Should it be `...` (three separate dots)?

File: ios/brave-ios/Sources/BrowserMenu/Identifiers.swift [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+import BraveWallet
+import Foundation
+
+extension Action.Identifier {
+
+  public static let vpn: Self = .init(
+    id: "ToggleVPN",
+    title: Strings.ActionTitles.vpn,
+    braveSystemImage: "leo.product.vpn",
+    defaultRank: 100,
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/BrowserMenu/Identifiers.swift:15]: What are these numbers and where do they come from? 
They don't seem to be sequential from `1...10` for example. They jump from 100 to 1600 to 1700, etc.
We should probably just add a comment explaining the rank so if someone adds a new item, they understand it :)

File: ios/brave-ios/Sources/BrowserMenu/VPNStatus.swift [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  var displayName: String
+
+  init(countryCode: String, displayName: String) {
+    self.flag = Self.flagEmojiForCountryCode(code: countryCode)
+    self.displayName = displayName
+  }
+
+  private static func flagEmojiForCountryCode(code: String) -> String {
+    // Root Unicode flags index
+    let rootIndex: UInt32 = 127397
```
Comment (Brandon-T) [inline ios/brave-ios/Sources/BrowserMenu/VPNStatus.swift:32]: Add a comment for this `Regional indicator symbol` offset.

File: ios/browser/ui/browser_menu/features.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_IOS_BROWSER_UI_BROWSER_MENU_FEATURES_H_
+#define BRAVE_IOS_BROWSER_UI_BROWSER_MENU_FEATURES_H_
+
+#import "base/feature_list.h"
+
+namespace brave::features {
```
Comment (Brandon-T) [inline ios/browser/ui/browser_menu/features.h:11]: nit: Not sure if we do this anywhere else as it's C++17 iirc. I'm all for it, but usually I think we do nested namespaces?

```c++
namespace brave {
namespace features {

  //...........
  
}  //  features
}  // brave
```

Ignore if I'm incorrect.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift (hunk 2 above)
Comment (kylehickinson) [reply ios/brave-ios/Sources/BrowserMenu/BrowserMenuStrings.swift:65]: Yep thats on purpose, specifically want to use the ellipsis character in this scenario (`…` vs `...`)

File: ios/brave-ios/Sources/BrowserMenu/Identifiers.swift (hunk 4 above)
Comment (kylehickinson) [reply ios/brave-ios/Sources/BrowserMenu/Identifiers.swift:15]: Good call, they're coming from a Figma list at the moment. They're sequential in 100s at the moment, specifically with enough gap that lets us add more items in between in the future. It was basically either I order the actual Identifiers by rank or by type and I chose type (feature links > page actions > activity share sheet actions). I'll add more information on this at the top of this file

File: ios/brave-ios/Sources/BrowserMenu/VPNStatus.swift (hunk 5 above)
Comment (kylehickinson) [reply ios/brave-ios/Sources/BrowserMenu/VPNStatus.swift:32]: I believe I just copied this from the original implementation but will update the comment

File: ios/browser/ui/browser_menu/features.h (hunk 6 above)
Comment (kylehickinson) [reply ios/browser/ui/browser_menu/features.h:11]: I've seen it done both ways in core, so I just picked this one (e.g. https://github.com/brave/brave-core/blob/1795aa0b6d519dc9806232e89c285ca6b15d13b3/components/ai_chat/core/common/features.h#L15), though after I wrote this originally it seems like we did have a discussion about nested namespaces in general https://bravesoftware.slack.com/archives/C7VLGSR55/p1732569420404999

It seems like the general consensus is that this is allowed for feature flags but can always remove it if we decide differently later

**Review by Brandon-T — APPROVED** [review]

**Pull request merged**